src/blr/cli/tagbam.py

```diff
@@ -1,58 +1,100 @@
+                    divider = read.query_name[match.start() - 1]
+                    read.query_name = read.query_name.replace(divider + full_tag_string, "")
+                    if not args.only_remove:
+                        read.set_tag(tag, match.group("value"), value_type=match.group("type"))
+                    summary[f"reads with tag {tag}"] += 1
+                else:
+                    summary[f"reads without tag {tag}"] += 1
 
             out.write(read)
 
-    logger.info(f"Alignments missing barcodes: {alignments_missing_bc}")
+    print_stats(summary, name="stats")
     logger.info("Finished")
 
 
+def build_regex_sam_tag(bam_tag, allowed_value_chars="ATGCN"):
+    """
+    Buidls regex string for SAM tags.
+    :param bam_tag: str, SAM tag to search for, e.g. BX
+    :param allowed_value_chars: str, characters allowed in SAM value
+    :return: str, regex expression of a SAM tag
+    """
+
+    # Build regex pattern strings
+    pattern_tag_value = f"[{allowed_value_chars}]+"
+    pattern_tag_types = f"[{ALLOWED_SAM_TAG_TYPES}]"
+
+    # Add strings to name match object variables to match.group(<name>)
+    regex_bam_tag = add_regex_name(pattern=bam_tag, name="tag")
+    regex_allowed_types = add_regex_name(pattern=pattern_tag_types, name="type")
+    regex_tag_value = add_regex_name(pattern=pattern_tag_value, name="value")
+
+    # regex pattern search for tag:type:value
+    regex_string = r":".join([regex_bam_tag, regex_allowed_types, regex_tag_value])
+    return regex_string
+
+
+def add_regex_name(pattern, name):
+    """
+    Formats a string to fit the regex pattern for getting named match objects groups
+    :param pattern: str, regex pattern
+    :param name: str, name of the match group
+    :return: str, named regex string
+    """
+    prefix = "(?P<"
+    infix = ">"
+    suffix = ")"
+    return prefix + name + infix + pattern + suffix
+
+
 def add_arguments(parser):
     parser.add_argument("input",
-                        help="SAM/BAM file with mapped reads which is to be tagged with barcode information. To read "
-                             "from stdin use '-'.")
+                        help="BAM file with SAM tag info in header. To read from stdin use '-'.")
 
     parser.add_argument("-o", "--output", default="-",
                         help="Write output BAM to file rather then stdout.")
-    parser.add_argument("-b", "--barcode-tag", default="BX",
-                        help="SAM tag for storing the error corrected barcode. Default: %(default)s")
-    parser.add_argument("-s", "--sequence-tag", default="RX",
-                        help="SAM tag for storing the uncorrected barcode sequence. Default: %(default)s")
+    parser.add_argument("-t", "--tags", default=["BX"], nargs="*", help="List of SAM tags. Default: %(default)s")
+    parser.add_argument("--only-remove", action="store_true", help="Only remove tag from header, will set SAM tag.")
```

Comment:
What do expect the use of this to be? Is there anywhere we would need only to remove from the tags?

Reply:
I understand it as `ema` takes input headers like this `@header:<BC SEQ> <BC TAG>` and outputs query headers like this `header:<BC SEQ>` with the `<BC TAG>` info is already set to a tag after mapping.

In that case one should write a new ema pattern and run it with the `--only-remove` flag to get rid of the duplicated information in the header.

The same goes for `bwa mem` if we want to use the `-H` option to tag the BAM file while mapping since it doesn't remove the header info either.